### PR TITLE
API added for history retrieve 

### PR DESCRIPTION
**Pull request by tasawernawaz**
[ECOM-7043](https://openedx.atlassian.net/browse/ECOM-7043)

### CODE REVIEW

**Comment by tasawernawaz** [comment]
@waheedahmed @awais786 Please review this.

File: course_discovery/apps/publisher/api/urls.py [hunk 1]
```diff
@@ -2,12 +2,13 @@
 from django.conf.urls import url
 
 from course_discovery.apps.publisher.api.views import (
-    CourseRoleAssignmentView, OrganizationGroupUserView, UpdateCourseKeyView
-)
+    CourseRoleAssignmentView, OrganizationGroupUserView, UpdateCourseKeyView,
+    CourseHistoryDetailView)
```
Comment (waheedahmed) [inline course_discovery/apps/publisher/api/urls.py]: Move closing `)` on next line and `CourseHistoryDetailView` fits onto the line above.

File: course_discovery/apps/publisher/api/tests/test_serializers.py [hunk 2]
```diff
@@ -4,9 +4,9 @@
 
 from course_discovery.apps.core.tests.factories import UserFactory
 from course_discovery.apps.publisher.api.serializers import (
-    CourseUserRoleSerializer, GroupUserSerializer, UpdateCourseKeySerializer
-)
-from course_discovery.apps.publisher.tests.factories import CourseUserRoleFactory, CourseRunFactory
+    CourseUserRoleSerializer, GroupUserSerializer, UpdateCourseKeySerializer,
+    CourseHistorySerializer)
```
Comment (waheedahmed) [inline course_discovery/apps/publisher/api/tests/test_serializers.py]: Same here and `CourseHistorySerializer` fits onto the line above.

File: course_discovery/apps/publisher/api/views.py [hunk 3]
```diff
@@ -1,12 +1,12 @@
...
 
 from course_discovery.apps.core.models import User
-from course_discovery.apps.publisher.models import CourseUserRole, OrganizationExtension, CourseRun
+from course_discovery.apps.publisher.models import CourseUserRole, OrganizationExtension, CourseRun, Course
 from course_discovery.apps.publisher.api.permissions import CanViewAssociatedCourse, InternalUserPermission
 from course_discovery.apps.publisher.api.serializers import (
-    CourseUserRoleSerializer, GroupUserSerializer, UpdateCourseKeySerializer
-)
+    CourseUserRoleSerializer, GroupUserSerializer, UpdateCourseKeySerializer,
+    CourseHistorySerializer)
```
Comment (waheedahmed) [inline course_discovery/apps/publisher/api/views.py]: Same here and `CourseHistorySerializer` fits onto the line above.

**Review by waheedahmed — CHANGES_REQUESTED** [review]

**Comment by tasawernawaz** [comment]
@waheedahmed Feedback addressed.

File: course_discovery/apps/publisher/api/tests/test_views.py [hunk 4]
```diff
@@ -323,3 +323,53 @@ def assert_email_sent(self, object_path, subject, expected_body):
...
+            'full_description': history.full_description,
+            'expected_learnings': history.expected_learnings,
+            'prerequisites': history.prerequisites,
+            'primary_subject': history.primary_subject.name,
+            'secondary_subject': history.secondary_subject.name,
+            'tertiary_subject': history.tertiary_subject.name
+        }
+
+        response = self._get_response(history.history_id)
+        self.assertEqual(json.loads(response.content.decode('utf-8')), expected)
```
Comment (waheedahmed) [inline course_discovery/apps/publisher/api/tests/test_views.py]: Instead of loading the content you can simply use `response.data` dict.

**Review by waheedahmed — CHANGES_REQUESTED** [review]

**Comment by tasawernawaz** [comment]
@bderusha Please review.

**Review by waheedahmed — APPROVED** [review]

**Comment by bderusha** [comment]
Why are we creating this endpoint rather than just exposing the history to the CourseDetailsView view that needs it?

This code looks fine, but it seems like unnecessary complexity

**Comment by awais786** [comment]
@bderusha we discussed that here. But we were thinking if a course has 20-30 revisions. Then response will become too large. That's why added a separate end-point.

**Comment by tasawernawaz** [comment]
@bderusha We need to show diff on selection of revision from dropdown, so if user selects any revision number we need to have that history object to compare it with current. We can't expose who history to view. We can expose most recent history object to view but I think it will be unnecessary to expose whole history at once. Your thoughts ?

**Comment by bderusha** [comment]
No, that makes a lot of sense.  I wasn't really thinking about the fact that we're retrieving the whole object and not just the version being actively diff'd

This looks good then :)

**Comment by tasawernawaz** [comment]
@bderusha Thanks, Are you going to review the code or I can merge.

File: course_discovery/apps/publisher/api/urls.py [hunk 5]
```diff
@@ -2,12 +2,13 @@
...
-    CourseRoleAssignmentView, OrganizationGroupUserView, UpdateCourseKeyView
+    CourseRoleAssignmentView, OrganizationGroupUserView, UpdateCourseKeyView, CourseHistoryDetailView
 )
 
 urlpatterns = [
     url(r'^course_role_assignments/(?P<pk>\d+)/$', CourseRoleAssignmentView.as_view(), name='course_role_assignments'),
     url(r'^admins/organizations/(?P<pk>\d+)/users/$', OrganizationGroupUserView.as_view(),
         name='organization_group_users'),
     url(r'^course_runs/(?P<pk>\d+)/$', UpdateCourseKeyView.as_view(), name='update_course_key'),
+    url(r'^course_history/(?P<history_id>\d+)/$', CourseHistoryDetailView.as_view(), name='course_history'),
```
Comment (bderusha) [inline course_discovery/apps/publisher/api/urls.py]: Can we name this `course_ revisions` to be consistent with the other urls referencing the same objects

**Review by bderusha — CHANGES_REQUESTED** [review]
Everything looks good.  My only request is that we rename all of the references from `history` to `revision`

File: course_discovery/apps/publisher/api/serializers.py [hunk 6]
```diff
@@ -71,3 +72,29 @@ def update(self, instance, validated_data):
             send_email_for_studio_instance_created(instance)
 
         return instance
+
+
+class CourseHistorySerializer(serializers.ModelSerializer):
```
Comment (bderusha) [inline course_discovery/apps/publisher/api/serializers.py]: `CourseRevisionSerializer`

File: course_discovery/apps/publisher/api/tests/test_serializers.py [hunk 7]
```diff
@@ -4,9 +4,9 @@
 
 from course_discovery.apps.core.tests.factories import UserFactory
 from course_discovery.apps.publisher.api.serializers import (
-    CourseUserRoleSerializer, GroupUserSerializer, UpdateCourseKeySerializer
+    CourseUserRoleSerializer, GroupUserSerializer, UpdateCourseKeySerializer, CourseHistorySerializer
```
Comment (bderusha) [inline course_discovery/apps/publisher/api/tests/test_serializers.py]: `CourseRevisionSerializer`

File: course_discovery/apps/publisher/api/tests/test_serializers.py [hunk 8]
```diff
@@ -71,3 +71,55 @@ def test_validation_error(self):
         serializer = self.serializer_class(self.course_run)
         with self.assertRaises(ValidationError):
             serializer.validate(serializer.data)
+
+
+class CourseHistorySerializerTests(TestCase):
```
Comment (bderusha) [inline course_discovery/apps/publisher/api/tests/test_serializers.py]: `CourseRevisionSerializerTests`

File: course_discovery/apps/publisher/api/tests/test_serializers.py [hunk 9]
```diff
@@ -71,3 +71,55 @@ def test_validation_error(self):
...
+    def test_course_history_serializer(self):
```
Comment (bderusha) [inline course_discovery/apps/publisher/api/tests/test_serializers.py]: `test_course_revision_serializer`

File: course_discovery/apps/publisher/api/tests/test_serializers.py [hunk 10]
```diff
@@ -71,3 +71,55 @@ def test_validation_error(self):
...
+        """ Verify that CourseHistorySerializer serialize the course history object. """
```
Comment (bderusha) [inline course_discovery/apps/publisher/api/tests/test_serializers.py]: `CourseRevisionSerializer`

File: course_discovery/apps/publisher/api/tests/test_serializers.py [hunk 11]
```diff
@@ -71,3 +71,55 @@ def test_validation_error(self):
...
+
+        course = CourseFactory()
+        course.title = 'updated title'
+        course.save()
+        history = course.history.first()
+        serializer = CourseHistorySerializer(history)
```
Comment (bderusha) [inline course_discovery/apps/publisher/api/tests/test_serializers.py]: `CourseRevisionSerializer`

File: course_discovery/apps/publisher/api/tests/test_serializers.py [hunk 12]
```diff
@@ -71,3 +71,55 @@ def test_validation_error(self):
...
+
+
+class CourseHistorySerializerTests(TestCase):
+    def test_course_history_serializer(self):
+        """ Verify that CourseHistorySerializer serialize the course history object. """
+
+        course = CourseFactory()
+        course.title = 'updated title'
+        course.save()
+        history = course.history.first()
```
Comment (bderusha) [inline course_discovery/apps/publisher/api/tests/test_serializers.py]: `revision`

File: course_discovery/apps/publisher/api/tests/test_views.py [hunk 13]
```diff
@@ -323,3 +323,53 @@ def assert_email_sent(self, object_path, subject, expected_body):
         self.assertIn(expected_body, body)
         page_url = 'https://{host}{path}'.format(host=Site.objects.get_current().domain.strip('/'), path=object_path)
         self.assertIn(page_url, body)
+
+
+class CourseHistoryDetailViewTests(TestCase):
```
Comment (bderusha) [inline course_discovery/apps/publisher/api/tests/test_views.py]: `CourseRevisionDetailViewTests`

File: course_discovery/apps/publisher/api/tests/test_views.py [hunk 14]
```diff
@@ -323,3 +323,53 @@ def assert_email_sent(self, object_path, subject, expected_body):
...
+        self.course = factories.CourseFactory()
+        self.course.title = "updated title"
+        self.course.save()
+
+        self.user = UserFactory()
+        self.client.login(username=self.user.username, password=USER_PASSWORD)
+
+    def test_get_course_history(self):
+        """Verify that api returns history object against given history id. """
+        history = self.course.history.first()
```
Comment (bderusha) [inline course_discovery/apps/publisher/api/tests/test_views.py]: `revision`

File: course_discovery/apps/publisher/api/urls.py [hunk 15]
```diff
@@ -2,12 +2,13 @@
...
-    CourseRoleAssignmentView, OrganizationGroupUserView, UpdateCourseKeyView
+    CourseRoleAssignmentView, OrganizationGroupUserView, UpdateCourseKeyView, CourseHistoryDetailView
 )
 
 urlpatterns = [
     url(r'^course_role_assignments/(?P<pk>\d+)/$', CourseRoleAssignmentView.as_view(), name='course_role_assignments'),
     url(r'^admins/organizations/(?P<pk>\d+)/users/$', OrganizationGroupUserView.as_view(),
         name='organization_group_users'),
     url(r'^course_runs/(?P<pk>\d+)/$', UpdateCourseKeyView.as_view(), name='update_course_key'),
+    url(r'^course_revisions/(?P<history_id>\d+)/$', CourseHistoryDetailView.as_view(), name='course_history'),
```
Comment (bderusha) [inline course_discovery/apps/publisher/api/urls.py]: `CourseRevisionDetailView`

Comment (bderusha) [reply course_discovery/apps/publisher/api/urls.py]: `name=course_revisions`

File: course_discovery/apps/publisher/api/views.py [hunk 16]
```diff
@@ -29,3 +29,10 @@ class UpdateCourseKeyView(UpdateAPIView):
     permission_classes = (IsAuthenticated, InternalUserPermission,)
     queryset = CourseRun.objects.all()
     serializer_class = UpdateCourseKeySerializer
+
+
+class CourseHistoryDetailView(RetrieveAPIView):
```
Comment (bderusha) [inline course_discovery/apps/publisher/api/views.py]: `CourseRevisionDetailView`

File: course_discovery/apps/publisher/api/views.py [hunk 17]
```diff
@@ -29,3 +29,10 @@ class UpdateCourseKeyView(UpdateAPIView):
...
+    permission_classes = (IsAuthenticated, )
+    serializer_class = CourseHistorySerializer
+    queryset = Course.history.all()  # pylint: disable=no-member
+    lookup_field = 'history_id'
```
Comment (bderusha) [inline course_discovery/apps/publisher/api/views.py]: `revision_id`

**Review by bderusha — CHANGES_REQUESTED** [review]
I meant change the name to be consistently used throughout.  Sorry if that wasn't clear.  I highlighted a few places, but the review comments are not exhaustive.

**Comment by tasawernawaz** [comment]
@bderusha Feedback addressed.

**Review by bderusha — APPROVED** [review]

**Pull request merged**